stores/chain.go

```diff
@@ -0,0 +1,356 @@
+				MaturityHeight: e.MaturityHeight,
+				Height:         index.Height,
+				BlockID:        hash256(index.ID),
+			}).Error; err != nil {
+			return nil
+		}
+	}
+
+	// create events
+	for _, e := range events {
+		if err := u.tx.
+			Clauses(clause.OnConflict{
+				DoNothing: true,
+				Columns:   []clause.Column{{Name: "event_id"}},
+			}).
+			Create(&dbWalletEvent{
+				EventID:        hash256(e.ID),
+				Inflow:         currency(e.Inflow),
+				Outflow:        currency(e.Outflow),
+				Transaction:    e.Transaction,
+				MaturityHeight: e.MaturityHeight,
+				Source:         string(e.Source),
+				Timestamp:      e.Timestamp.Unix(),
+				Height:         e.Index.Height,
+				BlockID:        hash256(e.Index.ID),
+			}).Error; err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Commit commits the updates to the database.
+func (u *chainUpdateTx) Commit() error {
+	return u.tx.Commit().Error
+}
+
+// Rollback rolls back the transaction
+func (u *chainUpdateTx) Rollback() error {
+	return u.tx.Rollback().Error
+}
+
+// ContractState returns the state of a file contract.
+func (u *chainUpdateTx) ContractState(fcid types.FileContractID) (api.ContractState, error) {
+	var state contractState
+	err := u.tx.
+		Select("state").
+		Model(&dbContract{}).
+		Where("fcid = ?", fileContractID(fcid)).
+		Scan(&state).
+		Error
+
+	if err == gorm.ErrRecordNotFound {
+		err = u.tx.
+			Select("state").
+			Model(&dbArchivedContract{}).
+			Where("fcid = ?", fileContractID(fcid)).
+			Scan(&state).
+			Error
+	}
+
+	if err != nil {
+		return "", err
+	}
+	return api.ContractState(state.String()), nil
+}
+
+// RemoveSiacoinElements is called with all siacoin elements that were spent in
+// the update.
+func (u *chainUpdateTx) RemoveSiacoinElements(ids []types.SiacoinOutputID) error {
+	for _, id := range ids {
+		if err := u.tx.
+			Where("output_id", hash256(id)).
+			Delete(&dbWalletOutput{}).
+			Error; err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RevertIndex is called with the chain index that is being reverted. Any
+// transactions and siacoin elements that were created by the index should be
+// removed.
+func (u *chainUpdateTx) RevertIndex(index types.ChainIndex, removed, unspent []types.SiacoinElement) error {
+	// recreate unspent outputs
+	for _, e := range unspent {
+		if err := u.tx.
+			Clauses(clause.OnConflict{
+				DoNothing: true,
+				Columns:   []clause.Column{{Name: "output_id"}},
+			}).
+			Create(&dbWalletOutput{
+				OutputID:       hash256(e.ID),
+				LeafIndex:      e.StateElement.LeafIndex,
+				MerkleProof:    merkleProof{proof: e.StateElement.MerkleProof},
+				Value:          currency(e.SiacoinOutput.Value),
+				Address:        hash256(e.SiacoinOutput.Address),
+				MaturityHeight: e.MaturityHeight,
+				Height:         index.Height,
+				BlockID:        hash256(index.ID),
+			}).Error; err != nil {
+			return nil
+		}
+	}
+
+	// remove outputs created at the reverted index
+	for _, e := range removed {
```

Comment:
I believe Nate recently updated `walletd` to reference the index that created an output or event via foreign key which we should probably consider here too. That way we only need to delete the reverted index and cascade delete will take care of the rest.

Reply:
Hm, good idea I guess, I'll do that in a separate PR though if that's ok.